contracts/MechMarketplace.sol

```diff
@@ -381,132 +380,246 @@ contract MechMarketplace is IErrorsMarketplace {
-        if (priorityMech != msg.sender) {
-            // Within the defined response time only a chosen priority mech is able to deliver
-            if (block.timestamp > mechDelivery.responseTimeout) {
-                // Decrease priority mech karma as the mech did not deliver
-                IKarma(karma).changeMechKarma(priorityMech, -1);
-                // Revoke request from the priority mech
-                IMech(priorityMech).revokeRequest(requestId);
-            } else {
-                // Priority mech responseTimeout is still >= block.timestamp
-                revert PriorityMechResponseTimeout(mechDelivery.responseTimeout, block.timestamp);
+            // Check if request has been delivered
+            if (mechDelivery.deliveryMech != address(0)) {
+                continue;
             }
-        }
 
-        // Record the actual delivery mech
-        mechDelivery.deliveryMech = msg.sender;
+            // TODO continue instead of revert? This would let us just ignore incorrect requests without reverting the bulk
+            // TODO Additional thought: The price between posting requests and delivering could have changed, thus just continue such that mech cleans those request info on their side
+            // Check for actual mech delivery rate
+            requesterDeliveryRates[i] = mechDelivery.deliveryRate;
+            if (mechDeliveryRates[i] > requesterDeliveryRates[i]) {
+                revert Overflow(mechDeliveryRates[i], mechDelivery.deliveryRate);
+            }
+
+            // If delivery mech is different from the priority one
+            if (priorityMech != msg.sender) {
+                // Within the defined response time only a chosen priority mech is able to deliver
+                if (block.timestamp > mechDelivery.responseTimeout) {
+                    // Decrease priority mech karma as the mech did not deliver
+                    // Needs to stay atomic as each different priorityMech can be any address
+                    IKarma(karma).changeMechKarma(priorityMech, -1);
+                } else {
+                    // TODO continue instead of revert? This would let us just ignore incorrect requests without reverting the bulk
+                    // TODO Additional thought: revert as mech has just miscalculated time, does not mean it has to clear that request
+                    // Priority mech responseTimeout is still >= block.timestamp
+                    revert PriorityMechResponseTimeout(mechDelivery.responseTimeout, block.timestamp);
+                }
+            }
+
+            // Record the actual delivery mech
+            mechDelivery.deliveryMech = msg.sender;
+
+            // Record requester
+            requesters[i] = mechDelivery.requester;
+
+            // Increase the amount of requester delivered requests
+            mapDeliveryCounts[requesters[i]]++;
+
+            // Mark request as delivered
+            deliveredRequests[i] = true;
```

Comment:
This is a key array that's returned to the mech - it allows for it to see which requests were actually delivered and which ones need to be cleaned up as they are not going to be delivered.